qutip/core/qobj.py

```diff
@@ -81,11 +81,18 @@ def _require_equal_type(method):
     """
     @functools.wraps(method)
     def out(self, other):
+        if isinstance(other, Qobj):
```

Comment:
Sorry, I didn't notice this was meant to return `NotImplemented` instead of raising, so please replace the raises everywhere. I also intended each `if` to end with a return.

The difference between my suggestion and what is here now is that the three cases are clearly distinguished at the top level of the function and the attempt to coerce non-Qobj other into a Qobj is gone.